cmd/inspect.go

```diff
@@ -15,142 +16,179 @@ import (
-				return ExitCodeError
-			}
-		}
-		if !constraints.Check(tflint.Version) {
-			cli.formatter.Print(tflint.Issues{}, fmt.Errorf("Failed to satisfy version constraints; tflint-ruleset-%s requires %s, but TFLint version is %s", name, constraints, tflint.Version), cli.loader.Sources())
-			return ExitCodeError
-		}
-
-		if err := ruleset.ApplyGlobalConfig(config); err != nil {
-			cli.formatter.Print(tflint.Issues{}, fmt.Errorf("Failed to apply global config to `%s` plugin; %w", name, err), cli.loader.Sources())
-			return ExitCodeError
-		}
-		configSchema, err := ruleset.ConfigSchema()
-		if err != nil {
-			cli.formatter.Print(tflint.Issues{}, fmt.Errorf("Failed to fetch config schema from `%s` plugin; %w", name, err), cli.loader.Sources())
-			return ExitCodeError
-		}
-		content := &hclext.BodyContent{}
-		if plugin, exists := cfg.Plugins[name]; exists {
-			var diags hcl.Diagnostics
-			content, diags = plugin.Content(configSchema)
-			if diags.HasErrors() {
-				cli.formatter.Print(tflint.Issues{}, fmt.Errorf("Failed to parse `%s` plugin config; %w", name, diags), cli.loader.Sources())
-				return ExitCodeError
-			}
-		}
-		err = ruleset.ApplyConfig(content, cfg.Sources())
-		if err != nil {
-			cli.formatter.Print(tflint.Issues{}, fmt.Errorf("Failed to apply config to `%s` plugin; %w", name, err), cli.loader.Sources())
-			return ExitCodeError
-		}
+	rootRunner := runners[len(runners)-1]
 
-		rulesets = append(rulesets, ruleset)
+	// Launch plugin processes
+	rulesetPlugin, err := launchPlugins(cli.config)
+	if rulesetPlugin != nil {
+		defer rulesetPlugin.Clean()
 	}
-	if err := cfg.ValidateRules(rulesets...); err != nil {
-		cli.formatter.Print(tflint.Issues{}, fmt.Errorf("Failed to check rule config; %w", err), cli.loader.Sources())
-		return ExitCodeError
+	if err != nil {
+		return tflint.Issues{}, err
 	}
 
 	// Run inspection
 	for _, ruleset := range rulesetPlugin.RuleSets {
 		for _, runner := range runners {
 			err = ruleset.Check(plugin.NewGRPCServer(runner, rootRunner, cli.loader.Files()))
 			if err != nil {
-				cli.formatter.Print(tflint.Issues{}, fmt.Errorf("Failed to check ruleset; %w", err), cli.loader.Sources())
-				return ExitCodeError
+				return tflint.Issues{}, fmt.Errorf("Failed to check ruleset; %w", err)
 			}
 		}
 	}
 
-	issues := tflint.Issues{}
 	for _, runner := range runners {
 		issues = append(issues, runner.LookupIssues(filterFiles...)...)
 	}
+	// Set module sources to CLI
+	for path, source := range cli.loader.Sources() {
+		cli.sources[path] = source
+	}
 
-	// Print issues
-	cli.formatter.Print(issues, nil, cli.loader.Sources())
+	return issues, nil
+}
 
-	if len(issues) > 0 && !cfg.Force {
-		return ExitCodeIssuesFound
+func (cli *CLI) withinChangedDir(dir string, proc func() error) error {
+	if dir != "." {
+		err := os.Chdir(dir)
+		if err != nil {
+			return fmt.Errorf("Failed to switch to a different working directory; %w", err)
+		}
 	}
 
-	return ExitCodeOK
+	if err := proc(); err != nil {
```

Comment:
If `proc()` fails, the working directory is never reset to its original value. It should probably be reset regardless of the inner result. Since modules are processed in series I don't think this would have any impact right now.

Reply:
Makes sense. Initially, I thought that if an error occurred, it should be returned to the caller without any changes, but It's obviously counterintuitive that the directory is changed outside of `withinChangedDir`.